src/context/common.ml

```diff
@@ -88,11 +96,22 @@ class compiler_callbacks = object(self)
 			List.iter (fun f -> try f() with Error.Error err -> handle_error err) (List.rev l);
 			self#run handle_error r
 
+	method run_with_types handle_error r (tl: module_type list) =
```

Comment:
This looks like it could be generalized to avoid the code duplication. The nature of the argument type isn't important here, so this should work with any argument type including `unit`.

Reply:
Done, but had to move it into a freestanding function as methods don't like polymorphism.